### PR TITLE
[NumericAxis Tests] - rework

### PR DESCRIPTION
**Pull request by bluong**
Completely reworked to adhere to current guidelines (using `.size()` instead of `[0].length`).  Also making use of `it`s inside of `forEach`'s to test against multiple orientations easily.

Test screenshot:
screen shot 2015-10-26 at 2 34 28 pm

### CODE REVIEW

File: test/axes/numericAxisTests.ts [hunk 1]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
+    });
 
-  it("uses the supplied Formatter", () => {
-    let SVG_WIDTH = 100;
-    let SVG_HEIGHT = 500;
-    let svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
-    let scale = new Plottable.Scales.Linear();
-    scale.range([0, SVG_HEIGHT]);
+    it("throws an error when setting an invalid position", () => {
+      assert.throws(() => axis.tickLabelPosition("top"), "horizontal");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: can you add assertion messages?

File: test/axes/numericAxisTests.ts [hunk 2]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
-                          .selectAll("." + Plottable.Axis.TICK_LABEL_CLASS)
-                          .filter(function(d: any, i: number) {
-                            return d3.select(this).style("visibility") === "visible";
-                          });
-    boundingBox = (<any> numericAxis)._element.select(".bounding-box").node().getBoundingClientRect();
-    visibleTickLabels[0].forEach((label: Element) => {
-      labelBox = label.getBoundingClientRect();
-      assertBoxInside(labelBox, boundingBox, 0, "long tick " + label.textContent + " is inside the bounding box");
+    it("throws an error when setting an invalid position", () => {
+      assert.throws(() => axis.tickLabelPosition("left"), "vertical");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: assertion message

File: test/axes/numericAxisTests.ts [hunk 3]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
+      let scale = new Plottable.Scales.Linear().domain([400000000, 500000000]);
+      let axis = new Plottable.Axes.Numeric(scale, "left");
 
-    let scale = new Plottable.Scales.Linear();
-    scale.domain([3, 0]);
+      axis.renderTo(svg);
 
-    let baseAxis = new Plottable.Axes.Numeric(scale, "bottom");
-    baseAxis.renderTo(svg);
+      let labelContainer = axis.content().select(".tick-label-container");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: factor out `tick-label-container`

File: test/axes/numericAxisTests.ts [hunk 4]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
-    baseAxis.renderTo(svg);
-    let tickMarks = (<any> baseAxis)._element.selectAll(".tick-mark");
-    tickMarks.each(function() {
-      let tickMark = d3.select(this);
-      let tickMarkPosition = Number(tickMark.attr("x"));
-      assert.isTrue(tickMarkPosition >= 0 && tickMarkPosition <= SVG_WIDTH, "tick marks are located within the bounding SVG");
+      let visibleTickLabels = numericAxis.content()
+        .selectAll(`.${Plottable.Axis.TICK_LABEL_CLASS}`)
+        .filter(function() {
+          return d3.select(this).style("visibility") === "visible";
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: what about `inherit`?

File: test/axes/numericAxisTests.ts [hunk 5]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
+
+    it("renders numbers in order with a reversed domain", () => {
+      let scale = new Plottable.Scales.Linear();
+      scale.domain([3, 0]);
+
+      let axis = new Plottable.Axes.Numeric(scale, "bottom");
+      axis.renderTo(svg);
+
+      let tickLabels = axis.content().selectAll(".tick-label")
+          .filter(function(d: any, i: number) {
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: can you factor this out? I've seen this filter a couple times

File: test/axes/numericAxisTests.ts [hunk 6]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
+
+      svg.remove();
+    });
   });
 
-  it("constrained tick labels do not overlap tick marks", () => {
-    let svg = TestMethods.generateSVG(100, 50);
+  describe("drawing tick labels when left oriented", () => {
+    it("draws ticks labels centered with the corresponding tick mark", () => {
+      let svg = TestMethods.generateSVG();
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: there's a couple duplicated lines, can you move them to beforeEach?
(can be consistent with `numericAxis` or `axis`)

File: test/axes/numericAxisTests.ts [hunk 7]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
+      assert.operator(tickLabels.size(), ">=", 2, "at least two tick labels were drawn");
+      let tickMarks = numericAxis.content().selectAll(`.${Plottable.Axis.TICK_MARK_CLASS}`);
+      assert.strictEqual(tickLabels.size(), tickMarks.size(), "there is one label per mark");
+
+      tickLabels.each(function(d, i) {
+        let tickLabelClientRect = this.getBoundingClientRect();
+        let tickMarkClientRect = (<Element> tickMarks[0][i]).getBoundingClientRect();
+        let labelCenter = (tickLabelClientRect.top + tickLabelClientRect.bottom) / 2;
+        let markCenter = (tickMarkClientRect.top + tickMarkClientRect.bottom) / 2;
+        assert.closeTo(labelCenter, markCenter, 1.5, `tick label ${i} is centered on mark`);
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: why 1.5?

File: test/axes/numericAxisTests.ts [hunk 8]
```diff
@@ -16,453 +16,438 @@ describe("NumericAxis", () => {
...
+      let scale = new Plottable.Scales.Linear();
+      scale.domain([5, -5]);
 
-    let svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+      let formatter = (d: any) => (d === 0) ? "ZERO" : String(d);
 
-    let testCases = [[-1, 1],
-                     [0, 10],
-                     [0, 999999999]];
+      let axis = new Plottable.Axes.Numeric(scale, "left");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: same here (duplicated code, and axis)

File: test/axes/numericAxisTests.ts (hunk 7 above)
Comment (bluong) [inline test/axes/numericAxisTests.ts]: The closeToRequirement was not enough

File: test/axes/numericAxisTests.ts [hunk 9]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
+  const isHorizontalOrientation = (orientation: string) => horizontalOrientations.indexOf(orientation) >= 0;
 
-    svg.remove();
+  horizontalOrientations.forEach((orientation) => {
+    const verticalTickLabelPositions = ["top", "bottom"];
+    it(`throws an error when setting an invalid position on orientation ${orientation}`, () => {
+      const scale = new Plottable.Scales.Linear();
+      const axis = new Plottable.Axes.Numeric(scale, orientation);
+      verticalTickLabelPositions.forEach((position) => {
+        (<any> assert).throws(() => axis.tickLabelPosition(position), "horizontal", "cannot set position");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: `cannot set position to ${position}`

File: test/axes/numericAxisTests.ts [hunk 10]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
-      labelBB = tickLabels[0][i].getBoundingClientRect();
-      assert.operator(markBB.bottom, "<=", labelBB.top, "tick label is below mark");
-    }
+  verticalOrientations.forEach((orientation) => {
+    const horizontalTickLabelPositions = ["left", "right"];
+    it(`throws an error when setting an invalid position on orientation ${orientation}`, () => {
+      const scale = new Plottable.Scales.Linear();
+      const axis = new Plottable.Axes.Numeric(scale, orientation);
+      horizontalTickLabelPositions.forEach((position) => {
+        (<any> assert).throws(() => axis.tickLabelPosition(position), "vertical", "cannot set position");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: \`cannot set position to ${position}`

File: test/axes/numericAxisTests.ts [hunk 11]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
+      });
+    });
+  });
 
-    svg.remove();
+  orientations.forEach((orientation) => {
+    const verticalTickLabelPositions = ["top", "bottom"];
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: can this be declared in outer scope?  it seems be used a couple times

File: test/axes/numericAxisTests.ts [hunk 12]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
-      let formattedValue = formatter(d);
-      assert.strictEqual(labelText, formattedValue, "The supplied Formatter was used to format the tick label");
+      svg.remove();
     });
 
-    svg.remove();
-  });
+    it("ensures that long labels are contained", () => {
+      let scale = new Plottable.Scales.Linear().domain([400000000, 500000000]);
+      let axis = new Plottable.Axes.Numeric(scale, "left");
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: why "left" isn't this block for bottom oriented axes?
also, maybe more axis declaration up a level and be consistent variable choice (axis vs numericAxis)

**Comment by ztsai** [comment]
can you also add an outer layer for `Axes` like how we have `Plots`?

File: test/axes/numericAxisTests.ts [hunk 13]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
-    let SVG_HEIGHT = 100;
-    let svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
+  it("does not overlap tick marks with tick labels", () => {
+    let svg = TestMethods.generateSVG();
+
     let scale = new Plottable.Scales.Linear();
-    let axis = new Plottable.Axes.Numeric(scale, "bottom");
-    axis.formatter((d: any) => "longstringsareverylong");
+    scale.domain([175, 185]);
+    let axis = new Plottable.Axes.Numeric(scale, "left")
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: numericAxis

File: test/axes/numericAxisTests.ts [hunk 14]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
+    });
   });
 
-  it("uses the supplied Formatter", () => {
-    let SVG_WIDTH = 100;
-    let SVG_HEIGHT = 500;
-    let svg = TestMethods.generateSVG(SVG_WIDTH, SVG_HEIGHT);
-    let scale = new Plottable.Scales.Linear();
-    scale.range([0, SVG_HEIGHT]);
+  describe("drawing tick labels when bottom oriented", () => {
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: why bottom only?

File: test/axes/numericAxisTests.ts [hunk 15]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
-    baseAxis.renderTo(svg);
+      let formatter = (d: any) => (d === 0) ? "ZERO" : String(d);
 
-    let tickLabels = (<any> baseAxis)._element.selectAll(".tick-label")
-        .filter(function(d: any, i: number) {
-          let visibility = d3.select(this).style("visibility");
-          return (visibility === "visible") || (visibility === "inherit");
-        });
-    assert.isTrue(tickLabels[0].length > 1, "more than one tick label is shown");
+      let axis = new Plottable.Axes.Numeric(scale, orientation);
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: numericAxis

File: test/axes/numericAxisTests.ts [hunk 16]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
+      axis.formatter(formatter);
+      axis.renderTo(svg);
 
-    for (let i = 0; i < tickLabels[0].length - 1; i++) {
-      let currLabel = d3.select(tickLabels[0][i]);
-      let nextLabel = d3.select(tickLabels[0][i + 1]);
-      assert.isTrue(Number(currLabel.text()) > Number(nextLabel.text()), "numbers are arranged in descending order from left to right");
-    }
+      let visibleTickLabels = applyVisibleFilter(axis.content()
+        .selectAll(`.${Plottable.Axis.TICK_LABEL_CLASS}`));
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: does this need to be on its own line?

**Comment by ztsai** [comment]
also, I am a bit concerned by the way `boxIsInside` is implemented, the inner box could be outside of the outer box by epsilon and still considered in

File: test/axes/numericAxisTests.ts [hunk 17]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
+      numericAxis.renderTo(svg);
 
-    let visibleTickLabels = (<any> baseAxis)._element.selectAll(".tick-label")
-      .filter(function(d: any, i: number) {
-        let visibility = d3.select(this).style("visibility");
-        return (visibility === "visible") || (visibility === "inherit");
+      let tickLabels = numericAxis.content().selectAll(`.${Plottable.Axis.TICK_LABEL_CLASS}`);
+      tickLabels.each(function(d, i) {
+        let labelText = d3.select(this).text();
+        let formattedValue = formatter(d);
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: inline?

File: test/axes/numericAxisTests.ts [hunk 18]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
+      let formatter = (d: any) => (d === 0) ? "ZERO" : String(d);
 
-    let tickLabels = (<any> baseAxis)._element.selectAll(".tick-label")
-        .filter(function(d: any, i: number) {
-          let visibility = d3.select(this).style("visibility");
-          return (visibility === "visible") || (visibility === "inherit");
-        });
-    assert.isTrue(tickLabels[0].length > 1, "more than one tick label is shown");
+      let axis = new Plottable.Axes.Numeric(scale, orientation);
+      axis.formatter(formatter);
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: why is this needed?

File: test/axes/numericAxisTests.ts [hunk 19]
```diff
@@ -16,453 +16,335 @@ describe("NumericAxis", () => {
...
-        .filter(function(d: any, i: number) {
-          let visibility = d3.select(this).style("visibility");
-          return (visibility === "visible") || (visibility === "inherit");
-        });
+      let formatter = (d: any) => (d === 0) ? "ZERO" : String(d);
 
-    tickLabels.each(function() {
-      let tickLabelRect = this.getBoundingClientRect();
+      let axis = new Plottable.Axes.Numeric(scale, verticalOrientation);
+      axis.formatter(formatter);
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: why is this needed

File: test/axes/numericAxisTests.ts [hunk 20]
```diff
@@ -1,468 +1,343 @@
...
 
-    let tickLabels = (<any> yAxis)._element.selectAll("." + Plottable.Axis.TICK_LABEL_CLASS)
-        .filter(function(d: any, i: number) {
-          let visibility = d3.select(this).style("visibility");
-          return (visibility === "visible") || (visibility === "inherit");
+        let visibleTickLabels = applyVisibleFilter(axis.content().selectAll(`.${Plottable.Axis.TICK_LABEL_CLASS}`));
+        let boundingBox = (<Element> svg.select(".axis").select(".bounding-box").node()).getBoundingClientRect();
+        visibleTickLabels.each(function(d, i) {
+          let visibleTickLabelRect = this.getBoundingClientRect();
+          assert.isTrue(boxIsInside(visibleTickLabelRect, boundingBox, 0.5), `tick label ${i} is inside the bounding box`);
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: why not use assertBoxInside

File: test/axes/numericAxisTests.ts [hunk 21]
```diff
@@ -1,468 +1,343 @@
...
-    let labelBB: ClientRect;
-    for (i = 0; i < tickLabels[0].length; i++) {
-      markBB = tickMarks[0][i].getBoundingClientRect();
-      let markCenter = (markBB.left + markBB.right) / 2;
-      labelBB = tickLabels[0][i].getBoundingClientRect();
-      let labelCenter = (labelBB.left + labelBB.right) / 2;
-      assert.closeTo(labelCenter, markCenter, 1, "tick label is centered on mark");
+describe("Axes", () => {
+  describe("NumericAxis", () => {
+    function boxIsInside(inner: ClientRect, outer: ClientRect, epsilon = 0) {
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: if this is only used in assertion, probably remove this and use `assertBoxInside` instead

File: test/axes/numericAxisTests.ts [hunk 22]
```diff
@@ -1,468 +1,343 @@
...
-      assert.strictEqual(labelText, formattedValue, "The supplied Formatter was used to format the tick label");
+        svg.remove();
+      });
     });
 
-    svg.remove();
-  });
+    orientations.forEach((orientation) => {
+      it(`separates tick labels with the same spacing for orientation ${orientation}`, () => {
+        let svg = TestMethods.generateSVG();
```
Comment (ztsai) [inline test/axes/numericAxisTests.ts]: you used `const` above and `let` here
why is that?

**Comment by ztsai** [comment]
can you group the tests directly under `NumericAxis` in describe blocks? that's a really big blob of tests

File: test/axes/numericAxisTests.ts (hunk 22 above)
Comment (bluong) [inline test/axes/numericAxisTests.ts]: Probably an inbetween of when we decided to use `const` over `let`
